File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickIMU;

type
  TExample = class
  private
    ipcon: TIPConnection;
    imu: TBrickIMU;
  public
    procedure QuaternionCB(sender: TBrickIMU; 
                           const x: single; const y: single;
                           const z: single; const w: single);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'a4JriVLwq7E'; { Change to your UID }

var
  e: TExample;

{ Quaternion callback }
procedure TExample.QuaternionCB(sender: TBrickIMU; 
                                const x: single; const y: single;
                                const z: single; const w: single);
begin
  WriteLn(Format('x: %.6f', [x]));
  WriteLn(Format('y: %.6f', [y]));
  WriteLn(Format('z: %.6f', [z]));
  WriteLn(Format('w: %.6f', [w]));
  WriteLn('');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  imu := TBrickIMU.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Set period for quaternion callback to 1s }
  imu.SetQuaternionPeriod(1000);

  { Register "quaternion callback" to procedure QuaternionCB }
  imu.OnQuaternion := {$ifdef FPC}@{$endif}QuaternionCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XXYYZZ" ' Change XXYYZZ to the UID of your IMU Brick

    ' Callback subroutine for quaternion callback
    Sub QuaternionCB(ByVal sender As BrickIMU, ByVal x As Single, ByVal y As Single, _
                     ByVal z As Single, ByVal w As Single)
        Console.WriteLine("Quaternion[X]: " + x.ToString())
        Console.WriteLine("Quaternion[Y]: " + y.ToString())
        Console.WriteLine("Quaternion[Z]: " + z.ToString())
        Console.WriteLine("Quaternion[W]: " + w.ToString())
        Console.WriteLine("")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim imu As New BrickIMU(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register quaternion callback to subroutine QuaternionCB
        AddHandler imu.QuaternionCallback, AddressOf QuaternionCB

        ' Set period for quaternion callback to 1s (1000ms)
        imu.SetQuaternionPeriod(1000)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
